flang/lib/Lower/OpenMP.cpp

```diff
@@ -1710,30 +1710,82 @@ bool ClauseProcessor::processLink(
 
 static mlir::omp::MapInfoOp
 createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
-                mlir::Value baseAddr, std::stringstream &name,
-                mlir::SmallVector<mlir::Value> bounds, uint64_t mapType,
-                mlir::omp::VariableCaptureKind mapCaptureType,
-                mlir::Type retTy) {
-  mlir::Value varPtr, varPtrPtr;
-  mlir::TypeAttr varType;
-
+                mlir::Value baseAddr, mlir::Value varPtrPtr, std::string name,
+                mlir::SmallVector<mlir::Value> bounds,
+                mlir::SmallVector<mlir::Value> members, uint64_t mapType,
+                mlir::omp::VariableCaptureKind mapCaptureType, mlir::Type retTy,
+                bool isVal = false) {
   if (auto boxTy = baseAddr.getType().dyn_cast<fir::BaseBoxType>()) {
     baseAddr = builder.create<fir::BoxAddrOp>(loc, baseAddr);
     retTy = baseAddr.getType();
   }
 
-  varPtr = baseAddr;
-  varType = mlir::TypeAttr::get(
+  mlir::TypeAttr varType = mlir::TypeAttr::get(
       llvm::cast<mlir::omp::PointerLikeType>(retTy).getElementType());
 
   mlir::omp::MapInfoOp op = builder.create<mlir::omp::MapInfoOp>(
-      loc, retTy, varPtr, varType, varPtrPtr, bounds,
+      loc, retTy, baseAddr, varType, varPtrPtr, members, bounds,
       builder.getIntegerAttr(builder.getIntegerType(64, false), mapType),
       builder.getAttr<mlir::omp::VariableCaptureKindAttr>(mapCaptureType),
-      builder.getStringAttr(name.str()));
+      builder.getStringAttr(name));
+
   return op;
 }
 
+static mlir::omp::MapInfoOp processDescriptorTypeMappings(
+    Fortran::semantics::SemanticsContext &semanticsContext,
+    Fortran::lower::StatementContext &stmtCtx,
+    Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+    mlir::Value descriptorAddr, mlir::Value descDataBaseAddr,
+    mlir::SmallVector<mlir::Value> &bounds, std::string asFortran,
+    llvm::omp::OpenMPOffloadMappingFlags mapCaptureType) {
+  llvm::SmallVector<mlir::Value> descriptorBaseAddrMembers;
+  fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+
+  mlir::Value descriptor = descriptorAddr;
+  if (fir::isAssumedShape(fir::unwrapRefType(descriptorAddr.getType()))) {
```

Comment:
It's necessary as these are not `fir.ref<fir.box<.,.>>` types like the allocatables and pointers, they're just `fir.box`'s. Which poses a little bit of a problem as the `fir::BoxOffsetOp` is actually only intended to work with `fir.ref<fir.box<.,.>>`'s. I asked @jeanPerier if it was possible to lower this restriction, but it is unfortunately not as there's some very problematic SSA IR you can generate if we allow non-ref boxes (a little beyond my current understanding). So, the way to use the `fir::BoxOffsetOp` to retrieve the base address in these cases appears to be to generate an intermediate alloca, store to it and then use this to access the required data. This is the method to convert to a `fir.ref<fir.box<.,.>>` that Jean mentioned, but perhaps there's another method to do so.  

I'll add a comment to describe this inline.

Reply:
Since you want to handle non-`!fir.ref` type here, the `fir::unwrapRefType` should not be used here.

Reply:
I think you should likely just go for `mlir::isa<fir::BaseBoxType>(descriptorAddr.getType())` because you will miss some case where descriptor are used outside of assumed shape.

An example is explicit shape polymorphic entities. We still pass that with a descriptor, but the extents in the fir.box type may be constant, so IsAssumedShape would return false.

I do not think these kinds of entities are likely to hit a OpenMP target region, but if the code here is meant to deal with fir.box, better directly test for it.

Reply:
Thank you, I'll do so, I may however, also add an assert or a TODO invocation to check for cases where something is not an AssumedShape/Pointer/Allocatable for the time being, as the other cases of box types are not intended to be covered by this PR just yet (they might work, but the PR has no tests for it and I've not expressly tested them locally like the other types, so I'd be doing a disservice saying they were supported I think).

Reply:
I opted to make the change to BaseBoxType, but not add the TODO emission for now, as making the swap but adding the TODO doesn't make too much sense in this case in hindsight.